src/cachelib/file.py

```diff
@@ -75,60 +70,85 @@ def _normalize_timeout(self, timeout):
         return int(timeout)
 
     def _list_dir(self):
-        """return a list of (fully qualified) cache filenames
-        """
-        mgmt_files = [self._get_filename(name).split('/')[-1]
-                      for name in (self._fs_count_file,)]
-        return [os.path.join(self._path, fn) for fn in os.listdir(self._path)
-                if not fn.endswith(self._fs_transaction_suffix)
-                and fn not in mgmt_files]
-
-    def _prune(self):
-        if self._threshold == 0 or not self._file_count > self._threshold:
-            return
-
+        """return a list of (fully qualified) cache filenames"""
+        mgmt_files = [
+            self._get_filename(name).split("/")[-1] for name in (self._fs_count_file,)
+        ]
+        return [
+            os.path.join(self._path, fn)
+            for fn in os.listdir(self._path)
+            if not fn.endswith(self._fs_transaction_suffix) and fn not in mgmt_files
+        ]
+
+    def _over_threshold(self):
+        return self._threshold != 0 and self._file_count > self._threshold
+
+    def _remove_expired(self, now):
         entries = self._list_dir()
-        now = time()
-        for idx, fname in enumerate(entries):
+        for fname in entries:
             try:
-                remove = False
-                with open(fname, 'rb') as f:
+                with open(fname, "rb") as f:
                     expires = pickle.load(f)
-                remove = (expires != 0 and expires <= now) or idx % 3 == 0
-
-                if remove:
+                if expires != 0 and expires < now:
                     os.remove(fname)
-            except (IOError, OSError):
+                    self._update_count(delta=-1)
+            except OSError:
                 pass
-        self._update_count(value=len(self._list_dir()))
+
+    def _remove_older(self):
+        entries = self._list_dir()
+        exp_fname_tuples = []
+        for fname in entries:
+            try:
+                with open(fname, "rb") as f:
+                    exp_fname_tuples.append((pickle.load(f), fname))
+            except OSError:
+                pass
+        fname_sorted = [
+            fname for _, fname in sorted(exp_fname_tuples, key=lambda item: item[1][0])
+        ]
+        for fname in fname_sorted:
```

Comment:
isn't this going to nuke the whole thing?

Reply:
Yep, nice catch. Fixed.